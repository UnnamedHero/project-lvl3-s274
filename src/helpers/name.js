import url from 'url';

export default (urlString) => {
  const { host, pathname } = url.parse(urlString);
  const validHost = host === null ? '' : host;
  const validPathName = pathname === '/' ? '' : pathname;
  const nonWordChars = new RegExp('\\W', 'g');
  const addrWithoutScheme = `${validHost}${validPathName}`;
  const replaceChar = '-';
  const name = addrWithoutScheme.replace(nonWordChars, replaceChar);
  return name;
};
