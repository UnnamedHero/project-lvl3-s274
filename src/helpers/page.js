import cheerio from 'cheerio';
import path from 'path';
import uniq from 'lodash/uniq';
import { makeResourceNameFromUrl } from './name';

const isAbsoluteUrl = link => link.search(new RegExp('(^(.+)?//)', 'g')) !== -1;

const resourceElements = [
  {
    name: 'link',
    attribute: 'href',
  },
  {
    name: 'script',
    attribute: 'src',
  },
  {
    name: 'img',
    attribute: 'src',
  },
];

export const makePage = html => cheerio.load(html);

export const getPageHtml = helper => helper.html();

export const getPageLinks = helper => resourceElements
  .reduce((acc, resource) => {
    const elementLinks = helper(resource.name)
      .map((_, elem) => cheerio(elem).attr(resource.attribute))
      .get();
    const uniqLinks = uniq(elementLinks);
    return [...acc, ...uniqLinks];
  }, []);

export const getPagaeLocalLinks = helper => getPageLinks(helper)
  .filter(link => !isAbsoluteUrl(link));

export const changePageLocalLinksTo = (helper, newPath) => {
  resourceElements.forEach((element) => {
    helper(element.name).each((_, elem) => {
      const link = cheerio(elem).attr(element.attribute);
      if (link && !isAbsoluteUrl(link)) {
        const newLink = makeResourceNameFromUrl(link);
        const fullPath = path.join(newPath, newLink);
        cheerio(elem).attr(element.attribute, fullPath);
      }
    });
  });
};
