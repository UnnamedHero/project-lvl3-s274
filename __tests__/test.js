import fs from 'fs-extra';
import os from 'os';
import nock from 'nock';
import path from 'path';
import pageLoader from '../src';
import { makeNameFromUrl } from '../src/helpers/name';

let tmpDir;
let htmlContent;

let innerTmpDir;
const existingFileName = 'foobar.html';

const testPath = __dirname;
const fixturesPath = path.join(testPath, '__fixtures__');
const pathTo = fileName => path.join(fixturesPath, fileName);

beforeAll(async () => {
  tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'page-loader'));
  innerTmpDir = await fs.mkdtemp(path.join(tmpDir, 'page-loader'));
  await fs.open(path.join(innerTmpDir, existingFileName), 'wx');
  const simpleHtml = 'simple.html';
  htmlContent = await fs.readFile(pathTo(simpleHtml), 'utf8');
});

afterAll(async () => {
  await fs.remove(tmpDir);
});

describe('page loader test', () => {
  const targetUrl = 'http://www.example.com';
  const name = makeNameFromUrl(targetUrl);
  const htmlName = `${name}.html`;
  const filesPath = `${name}_files`;

  const expectedDownloadedFiles = [
    { downloaded: '/icons-logo.png', fixture: 'logo.png' },
    { downloaded: '/css-style1.css', fixture: 'style1.css' },
    { downloaded: '/js-bitcoin-miner.js', fixture: 'bitcoin-miner.ts' },
    { downloaded: '/pictures-passport-1.png', fixture: 'passport-1.png' },
    { downloaded: '/scans-my-credit-card.jpg', fixture: 'my-credit-card.jpg' },
  ];

  test('should fail on fake dir', async () => {
    await expect(pageLoader('foobar', 'c:\\windows\\system32')).rejects.toMatchSnapshot();
  });

  test('should fail on readonly dir', async () => {
    await expect(pageLoader('foobar', '/')).rejects.toMatchSnapshot();
  });

  test('download page', async () => {
    nock(targetUrl)
      .get('/')
      .reply(200, htmlContent, { 'Content-Type': 'text/html' })
      .get('/icons/logo.png')
      .replyWithFile(200, pathTo('logo.png'))
      .get('/css/style1.css')
      .delay(1000)
      .replyWithFile(200, pathTo('style1.css'))
      .get('/js/bitcoin-miner.js')
      .delay(2000)
      .replyWithFile(200, pathTo('bitcoin-miner.ts'))
      .get('/pictures/passport-1.png')
      .replyWithFile(200, pathTo('passport-1.png'))
      .get('/scans/my-credit-card.jpg')
      .replyWithFile(200, pathTo('my-credit-card.jpg'));

    await pageLoader(targetUrl, tmpDir);
    const resultHtmlFilePath = path.join(tmpDir, htmlName);
    await expect(fs.open(resultHtmlFilePath, 'r')).resolves.toBeDefined();
  });

  test.each(expectedDownloadedFiles)('check if \'%o\' downloaded', async ({ downloaded, fixture }) => {
    const filePath = path.join(tmpDir, filesPath, downloaded);
    const expectedContent = await fs.readFile(pathTo(fixture), 'utf8');
    const downloadedContent = await fs.readFile(filePath, 'utf8');
    expect(downloadedContent).toEqual(expectedContent);
  });

  test('should fail if html file exists', async () => {
    const matcher = new RegExp(`^(EEXIST: file already exists, open '(.+)?${existingFileName}')$`, 'g');
    await expect(pageLoader('foobar', innerTmpDir))
      .rejects.toHaveProperty('message', expect.stringMatching(matcher));
  });

  test('should fail on 404', async () => {
    const url = 'https://geekbrains.ru/';
    nock(url)
      .get('/')
      .replyWithError(404);
    await expect(pageLoader(url, tmpDir))
      .rejects.toMatchSnapshot();
  });
});
