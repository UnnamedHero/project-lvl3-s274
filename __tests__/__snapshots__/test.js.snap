// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page loader test should fail on 404 1`] = `[Error: Error 404 for https://geekbrains.ru/]`;

exports[`page loader test should fail on fake dir 1`] = `[Error: ENOENT: no such file or directory, access 'c:\\windows\\system32']`;

exports[`page loader test should fail on readonly dir 1`] = `[Error: EACCES: permission denied, access '/']`;
